// metal-soy/test/assets/HelloWorld.soy.js
/* jshint ignore:start */
import Component from 'metal-component/src/Component';
import Soy from '../../src/Soy';
var templates;
goog.loadModule(function(exports) {

// This file was automatically generated from HelloWorld.soy.
// Please don't edit this file by hand.

/**
 * @fileoverview Templates in namespace HelloWorld.
 * @public
 */

goog.module('HelloWorld.incrementaldom');

/** @suppress {extraRequire} */
var soy = goog.require('soy');
/** @suppress {extraRequire} */
var soydata = goog.require('soydata');
/** @suppress {extraRequire} */
goog.require('goog.i18n.bidi');
/** @suppress {extraRequire} */
goog.require('goog.asserts');
var IncrementalDom = goog.require('incrementaldom');
var ie_open = IncrementalDom.elementOpen;
var ie_close = IncrementalDom.elementClose;
var ie_void = IncrementalDom.elementVoid;
var ie_open_start = IncrementalDom.elementOpenStart;
var ie_open_end = IncrementalDom.elementOpenEnd;
var itext = IncrementalDom.text;
var iattr = IncrementalDom.attr;


/**
 * @param {Object<string, *>=} opt_data
 * @param {(null|undefined)=} opt_ignored
 * @param {Object<string, *>=} opt_ijData
 * @return {void}
 * @suppress {checkTypes}
 */
function $render(opt_data, opt_ignored, opt_ijData) {
  opt_data = opt_data || {};
  ie_open('span', null, null,
      'class', 'render');
    $content(opt_data, null, opt_ijData);
  ie_close('span');
}
exports.render = $render;
if (goog.DEBUG) {
  $render.soyTemplateName = 'HelloWorld.render';
}


/**
 * @param {Object<string, *>=} opt_data
 * @param {(null|undefined)=} opt_ignored
 * @param {Object<string, *>=} opt_ijData
 * @return {void}
 * @suppress {checkTypes}
 */
function $content(opt_data, opt_ignored, opt_ijData) {
  opt_data = opt_data || {};
  ie_open('span', null, null,
      'class', 'content');
    itext('Hello ');
    itext((goog.asserts.assert((opt_data.name ? opt_data.name : 'World') != null), opt_data.name ? opt_data.name : 'World'));
    itext('!');
  ie_close('span');
}
exports.content = $content;
if (goog.DEBUG) {
  $content.soyTemplateName = 'HelloWorld.content';
}

exports.render.params = ["name"];
exports.render.types = {"name":"any"};
exports.content.params = ["name"];
exports.content.types = {"name":"any"};
templates = exports;
return exports;

});

class HelloWorld extends Component {}
Soy.register(HelloWorld, templates);
export default templates;
export { HelloWorld, templates };
/* jshint ignore:end */
