// metal-soy/test/assets/TemplateData.soy.js
/* jshint ignore:start */
import Component from 'metal-component/src/Component';
import Soy from '../../src/Soy';
var templates;
goog.loadModule(function(exports) {

// This file was automatically generated from TemplateData.soy.
// Please don't edit this file by hand.

/**
 * @fileoverview Templates in namespace TemplateData.
 * @public
 */

goog.module('TemplateData.incrementaldom');

/** @suppress {extraRequire} */
var soy = goog.require('soy');
/** @suppress {extraRequire} */
var soydata = goog.require('soydata');
/** @suppress {extraRequire} */
goog.require('goog.i18n.bidi');
/** @suppress {extraRequire} */
goog.require('goog.asserts');
var IncrementalDom = goog.require('incrementaldom');
var ie_open = IncrementalDom.elementOpen;
var ie_close = IncrementalDom.elementClose;
var ie_void = IncrementalDom.elementVoid;
var ie_open_start = IncrementalDom.elementOpenStart;
var ie_open_end = IncrementalDom.elementOpenEnd;
var itext = IncrementalDom.text;
var iattr = IncrementalDom.attr;


/**
 * @param {Object<string, *>=} opt_data
 * @param {(null|undefined)=} opt_ignored
 * @param {Object<string, *>=} opt_ijData
 * @return {void}
 * @suppress {checkTypes}
 */
function $render(opt_data, opt_ignored, opt_ijData) {
  opt_data = opt_data || {};
  ie_open('span');
    $subTemplate(opt_data, null, opt_ijData);
  ie_close('span');
}
exports.render = $render;
if (goog.DEBUG) {
  $render.soyTemplateName = 'TemplateData.render';
}


/**
 * @param {Object<string, *>=} opt_data
 * @param {(null|undefined)=} opt_ignored
 * @param {Object<string, *>=} opt_ijData
 * @return {void}
 * @suppress {checkTypes}
 */
function $subTemplate(opt_data, opt_ignored, opt_ijData) {
  opt_data = opt_data || {};
  ie_open('span', null, null,
      'class', 'subTemplate');
    itext((goog.asserts.assert((opt_data.foo) != null), opt_data.foo));
  ie_close('span');
}
exports.subTemplate = $subTemplate;
if (goog.DEBUG) {
  $subTemplate.soyTemplateName = 'TemplateData.subTemplate';
}

exports.render.params = [];
exports.render.types = {};
exports.subTemplate.params = ["foo"];
exports.subTemplate.types = {"foo":"any"};
templates = exports;
return exports;

});

class TemplateData extends Component {}
Soy.register(TemplateData, templates);
export default templates;
export { TemplateData, templates };
/* jshint ignore:end */
